openmeter/ent/schema/productcatalog.go

```diff
@@ -0,0 +1,251 @@
+			Optional().
+			Nillable(),
+		field.String("discounts").
+			GoType([]plan.Discount{}).
+			ValueScanner(DiscountsValueScanner).
+			SchemaType(map[string]string{
+				dialect.Postgres: "jsonb",
+			}).
+			Optional().
+			Nillable(),
+		field.String("plan_id").
+			NotEmpty().
+			Comment("The plan identifier the phase is assigned to."),
+	}
+}
+
+func (PlanPhase) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.From("plan", Plan.Type).
+			Ref("phases").
+			Field("plan_id").
+			Required().
+			Unique(),
+		edge.To("ratecards", PlanRateCard.Type).
+			Annotations(entsql.Annotation{
+				OnDelete: entsql.Cascade,
+			}),
+	}
+}
+
+func (PlanPhase) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "key"),
+		index.Fields("plan_id", "key").
+			Unique(),
+	}
+}
+
+type PlanRateCard struct {
+	ent.Schema
+}
+
+func (PlanRateCard) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.KeyMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+	}
+}
+
+type Price string
+
+func (PlanRateCard) Fields() []ent.Field {
+	return []ent.Field{
+		field.Enum("type").
+			GoType(plan.RateCardType("")).
+			Immutable(),
+		field.String("name").
+			NotEmpty(),
+		field.String("description").
+			Optional().
+			Nillable(),
+		field.String("feature_key").
+			Optional().
+			Nillable(),
+		field.String("entitlement_template").
+			GoType(&plan.EntitlementTemplate{}).
+			ValueScanner(EntitlementTemplateValueScanner).
+			SchemaType(map[string]string{
+				dialect.Postgres: "jsonb",
+			}).
+			Optional().
+			Nillable(),
+		field.String("tax_config").
+			GoType(&plan.TaxConfig{}).
+			ValueScanner(TaxConfigValueScanner).
+			SchemaType(map[string]string{
+				dialect.Postgres: "jsonb",
+			}).
+			Optional().
+			Nillable(),
+		field.String("billing_cadence").
+			GoType(datex.ISOString("")).
+			Optional().
+			Nillable(),
+		field.String("price").
+			GoType(&plan.Price{}).
+			ValueScanner(PriceValueScanner).
+			SchemaType(map[string]string{
+				dialect.Postgres: "jsonb",
+			}).
+			Optional().
+			Nillable(),
+		field.String("phase_id").
+			NotEmpty().
+			Comment("The phase identifier the ratecard is assigned to."),
+		field.String("feature_id").
+			Optional().
+			Nillable().
+			Comment("The feature identifier the ratecard is related to."),
+	}
+}
+
+func (PlanRateCard) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.From("phase", PlanPhase.Type).
+			Ref("ratecards").
+			Field("phase_id").
+			Required().
+			Unique(),
+		edge.From("features", Feature.Type).
+			Ref("ratecard").
+			Field("feature_id").
+			Unique(),
+	}
+}
+
+func (PlanRateCard) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "key"),
+		index.Fields("phase_id", "key").
+			Unique(),
+		index.Fields("phase_id", "feature_key").
+			Unique(),
+	}
+}
+
+func JSONStringValueScanner[T any]() field.ValueScannerFunc[T, *sql.NullString] {
```

Comment:
I'd suggest to move this under `pkg/framework`

Reply:
Done.